### PR TITLE
Feature/initial integration

### PR DESCRIPTION
**Pull request by prestonwrthompson**
***Settings***

**App Group Id**
The app group's API Identifier for Appboy.

**Update Existing Users Only**
Whether requests forwarded to Appboy from Segment should only update user's for whom the provided user IDs exist.

***Methods***

**Identify**

This will simply map the given attributes to the attributes section of the [Appboy user track endpoint](https://documentation.appboy.com/REST_APIs/User_Data#user-track).

**Track**

This will map the actions to the events section of the Appboy user track endpoint.

*Completed Order*

Track requests with the event name ```Completed Order``` according to the [Segment ECommerce API](https://segment.com/docs/api/tracking/ecommerce/) will be mapped to the purchases section of the Appboy user track endpoint.

**Page**

Page requests are currently not supported.

**Screen**

Screen requests are currently not supported.

**Group**

Group requests will set a custom attribute of 
```ruby
"ab_segmentio_group_" + groupId
```
 to true for the provided user through the attributes section of the Appboy user track endpoint. Group traits will not be tracked.

**Alias**

Alias requests are currently not supported.

### CODE REVIEW

File: lib/mapper.js [hunk 1]
```diff
@@ -8,8 +8,13 @@
...
  */
 
-exports.identify = function(identify){
-  return identify.json();
+exports.identify = function(identify, settings){
+  var traits = formatTraits(identify, settings);
+  return {
+    app_group_id: settings.appGroupId,
+    attributes: [traits],
+    src: "segment"
```
Comment (amillet89) [inline lib/mapper.js]: nitpick: we prefer single quotes for literals

File: lib/mapper.js [hunk 2]
```diff
@@ -21,6 +26,143 @@ exports.identify = function(identify){
  * @api private
  */
 
-exports.track = function(track){
-  return track.json();
+exports.track = function(track, settings){
+  var result = {
+    app_group_id: settings.appGroupId,
+    src: "segment"
```
Comment (amillet89) [inline lib/mapper.js]: same here

File: lib/mapper.js [hunk 3]
```diff
@@ -21,6 +26,143 @@ exports.identify = function(identify){
...
+  if (event && event.match(/completed ?order/i)) {
+    var purchases = [];
+    var products = track.products();
+
+    // add purchase for each product
+    products.forEach(function(product){
+      purchases.push({
+        external_id: track.userId(),
+        product_id: product.id,
+        currency: track.proxy('properties.currency') || "USD",
```
Comment (amillet89) [inline lib/mapper.js]: and here

File: lib/mapper.js [hunk 4]
```diff
@@ -21,6 +26,143 @@ exports.identify = function(identify){
...
+
+exports.group = function(group, settings){
+  // set basic attributes
+  var attributes = {
+    external_id: group.userId(),
+    _update_existing_only: settings.updateExistingOnly
+  };
+
+  // add group id as custom attribute key
+  var groupIdKey = "ab_segment_group_" + group.groupId();
```
Comment (amillet89) [inline lib/mapper.js]: and here

File: lib/mapper.js [hunk 5]
```diff
@@ -21,6 +26,143 @@ exports.identify = function(identify){
...
+  attributes[groupIdKey] = true;
+
+  return {
+    app_group_id: settings.appGroupId,
+    attributes: [attributes],
+    src: "segment"
```
Comment (amillet89) [inline lib/mapper.js]: here

File: lib/mapper.js [hunk 6]
```diff
@@ -21,6 +26,143 @@ exports.identify = function(identify){
...
-  return track.json();
+exports.track = function(track, settings){
+  var result = {
+    app_group_id: settings.appGroupId,
+    src: "segment"
+  };
+
+  var event = track.event();
+  // if the event is "Completed Order" from ECommerce API, then track as purchases instead of event
+  if (event && event.match(/completed ?order/i)) {
```
Comment (amillet89) [inline lib/mapper.js]: You actually don't need to do the event name matching here but can implement a separate completedOrder method, as in the Google Analytics implementation here: https://github.com/segmentio/integration-google-analytics/blob/master/lib/mapper.js#L75

**Comment by amillet89** [comment]
Could you squash the commits? Left a couple comments inline, otherwise looks great!

**Comment by prestonwrthompson** [comment]
Replied to all comments and squashed at #3! Thanks so much!

**Pull request closed without merging**